test/unit/org/apache/cassandra/index/sai/utils/RangeConcatIteratorTest.java

```diff
@@ -0,0 +1,443 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.util.function.Supplier;
+import java.util.stream.IntStream;
+
+import org.junit.Test;
+
+import static org.apache.cassandra.index.sai.utils.LongIterator.convert;
+import static org.apache.cassandra.index.sai.utils.RangeIterator.Builder.IteratorType.CONCAT;
+import static org.apache.cassandra.index.sai.utils.RangeIterator.Builder.IteratorType.INTERSECTION;
+import static org.apache.cassandra.index.sai.utils.RangeIterator.Builder.IteratorType.UNION;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class RangeConcatIteratorTest extends AbstractRangeIteratorTester
+{
+    @Test
+    public void testValidation()
+    {
+        try
+        {
+            buildConcat(build(1L, 4L), build(2L, 3L));
+            fail("Flows for a merging concatenation must not contain one another.");
+        }
+        catch (IllegalArgumentException ignored)
+        {
+        }
+
+        try
+        {
+            buildConcat(build(1L, 4L), build(2L, 5L));
+            fail("Minimum for flow must not be included in exclusive range of previous flow.");
+        }
+        catch (IllegalArgumentException ignored)
+        {
+        }
+
+        // allow min boundary included
+        RangeIterator concat = buildConcat(build(1L, 4L), build(4L, 5L));
+        assertEquals(convert(1L, 4L, 4L, 5L), convert(concat));
+
+        try
+        {
+            buildConcat(build(1L, 4L), build(0L, 3L));
+            fail("Maximum for flow must not be included in exclusive range of previous flow.");
+        }
+        catch (IllegalArgumentException ignored)
+        {
+        }
+
+        try
+        {
+            buildConcat(build(2L, 4L), build(0L, 1L));
+            fail("Flows for merging concatenation must be sorted.");
+        }
+        catch (IllegalArgumentException ignored)
+        {
+        }
```

Comment:
For all these validation tests, I was going to say we should remove the "flow" references, but why don't we just go a step farther and replace these w/ the...

```Assertions.assertThatThrownBy(() -> blah()).isInstanceOf(IllegalArgumentException.class).hasMessage("blah");```

...pattern? We could pull out a constant pattern for the error message in `Statistics#update()`, something like `ILLEGAL_SORTED_RANGE_MESSAGE`?